### PR TITLE
offline appenderator factory.

### PR DESCRIPTION
**Pull request by b-slim**
This patch adds an offline factory to create offline appenderator.

### CODE REVIEW

**Comment by b-slim** [comment]
@gianm can you please check this. I am not sure about the inject stuff if such class should be only created by Jackson or it is ok to have the `@Inject` annotation

File: server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    );
+
+    Appenderator appenderator = defaultOfflineAppenderatorFactory.build(
+        schema,
+        tuningConfig,
+        new FireDepartmentMetrics()
+    );
+    Assert.assertEquals(null, appenderator.startJob());
+    Assert.assertEquals("dataSourceName", appenderator.getDataSource());
+    appenderator.close();
```
Comment (leventov) [inline server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java]: Use try-with-resources instead of explicit `close()`

File: server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+}
```
Comment (leventov) [inline server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java]: Miss newline

File: server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        null
+    );
+
+    Appenderator appenderator = defaultOfflineAppenderatorFactory.build(
+        schema,
+        tuningConfig,
+        new FireDepartmentMetrics()
+    );
+    Assert.assertEquals(null, appenderator.startJob());
+    Assert.assertEquals("dataSourceName", appenderator.getDataSource());
```
Comment (leventov) [inline server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java]: Maybe try to add some sample row to Appenderator, to see that it works (and cache is non-null)?

File: server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java [hunk 4]
```diff
@@ -124,8 +124,8 @@ public boolean isAnnounced(DataSegment segment)
         null,
         indexIO,
         indexMerger,
-        null,
-        null
+        MapCache.create(500000),
```
Comment (leventov) [inline server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java]: Doesn't seem to be universal to use specifically `MapCache` (`CaffeineCache` is also an in-memory, local cache) of the specific size. Maybe replace `Appendirators` with something like `AppendiratorBuilder`, which allows to configure those batteries of arguments one-by-one using named configuration methods? It may distinguish between "realtime" and "offline" on different stages, e. g. in the beginning of the chain:
```java
AppenderatorBuilder.offline() // or .realtime(segmentAnnouncer)
  .setIndexMerger(...)
  .setCache(...)
  .build();
```

or in the end:
```java
AppenderatorBuilder.create()
  .setIndexMerger(..)
  .setCache(..)
  .buildRealtime() // or .buildOffline()
```

Or in some other way

Comment (leventov) [inline server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java]: Could you please also test this cache configuration missing, e. g. add something to test that would fail before this change? (Also see my comment to `DefaultOfflineAppenderatorFactoryTest`)

File: server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java (hunk 1 above)
Comment (b-slim) [inline server/src/test/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactoryTest.java]: ok

File: server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java (hunk 4 above)
Comment (b-slim) [inline server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java]: the test in this PR will fail if the cache was missing is that not enough ?

Comment (b-slim) [inline server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java]: @leventov thanks for the comments about the cache actually it is not really needed i will remove it.

**Comment by b-slim** [comment]
Thanks @leventov, i have addressed the comments minus the one about the builder. You think a Builder is really better ? if so i can change it. Please let me know.

**Comment by leventov** [comment]
@b-slim let some other Druid committer decide.

**Comment by b-slim** [comment]
@leventov thanks for the feedback !

**Comment by b-slim** [comment]
@gianm can you take look at this ? it is pretty minor

File: server/src/main/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactory.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+public class DefaultOfflineAppenderatorFactory implements AppenderatorFactory
+{
+  private final DataSegmentPusher dataSegmentPusher;
+  private final ObjectMapper objectMapper;
+  private final IndexIO indexIO;
+  private final IndexMerger indexMerger;
+
+  @Inject
```
Comment (cheddar) [inline server/src/main/java/io/druid/segment/realtime/appenderator/DefaultOfflineAppenderatorFactory.java]: Given that we expect it to be instantiated by Jackson, this would need to be `@JsonCreate`

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java [hunk 6]
```diff
@@ -912,7 +912,9 @@ public Object apply(@Nullable Object input)
                 identifier.getShardSpec().createChunk(sink)
             );
             for (FireHydrant hydrant : sink) {
-              cache.close(SinkQuerySegmentWalker.makeHydrantCacheIdentifier(hydrant));
+              if (cache != null) {
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: when can this be null ?

File: server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java [hunk 7]
```diff
@@ -129,5 +128,4 @@ public boolean isAnnounced(DataSegment segment)
     );
   }
 
-
```
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java]: unrelated change to this PR.

**Comment by cheddar** [comment]
Btw, you should be able to instantiate one of these locally on your dev box, which should make it clear if the injection/jackson stuff is working.

It should also be unit testable...

File: server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java (hunk 7 above)
Comment (b-slim) [inline server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java]: this is part of bug fix. when using offline appenderator we don't server queries so no need for cache.

**Comment by b-slim** [comment]
@cheddar thanks i guess the new UT fix that.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java (hunk 6 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: We should also remove the null check from the constr also then?

File: server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java (hunk 7 above)
Comment (b-slim) [inline server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java]: it is just less empty line is that not ok ?

Comment (nishantmonu51) [inline server/src/main/java/io/druid/segment/realtime/appenderator/Appenderators.java]: it's ok.

**Comment by b-slim** [comment]
@cheddar i have addressed all the comments Thanks !

**Comment by cheddar** [comment]
👍
